doc/acl/Dynamic-ACL-Design.md

```diff
@@ -0,0 +1,190 @@
+                                        "destination-ip-address":"192.168.0.3/32"
+                                    }
+                                },
+                                "ttl":"300"
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+```
+The YANG of ACL_RULE is required to be updated to accept new field `is_dynamic`
+
+Orchagent (actually `aclorch`) won't consume the value of `is_dynamic`. So no change is required to `orchagent`.
+#### STATE DB
+A new table is introduced to state_db to record the timestamp of creation, expiration and TTL of an ACL rule.
+```
+  key                      = ACL_TTL_TABLE:acl_rule_name     ; acl_rule_name specifies the corresponding ACL rule name, must be unique
+  ;field                   = value  
+  creation_time            = Integer                         ; timestamp when the rule is created or refreshed  
+  expiration_time          = Integer                         ; timestamp when the rule will expire
+  ttl                      = Integer                         ; the ttl, in second 
+```
+A sample
+```json
+{
+    "ACL_TTL_TABLE|DATAACL|RULE_1":{
+        "creation_time":1645686213,
+        "expiration_time":1645686513,
+        "ttl":300
+    }
+}
+```
+#### acl-loader
+
+Update `acl-loader` script to parse new field `is_dynamic`. The entry will be created in `ACL_TTL_TABLE` in `state_db` if `ttl` is present for certain an ACL rule in `acl.json`. Please find more details in workflow diagram.  
+
+#### acl_ttl_checker
+
+A helper script will be added to `swss` container. The checker is started after `orchagent` and check the TTL of dynamic ACL rules every 10 seconds by default. It will walk through all entries in `ACL_TTL_TABLE` and delete the corresponding `ACL_RULE` from `config_db` if
+
+- Current timestamp is larger than the `expiration_time` in the entry
+- The corresponding ACL rule is marked as dynamic
+
+### Work flow
+#### Add dynamic ACL rule
+<p align=center>
+<img src="img/dynamic_acl_creation.png" alt="Figure 1. Create dynamic ACL rule workflow">
+</p>
+
+#### Refresh TTL of existing ACL rule
+<p align=center>
+<img src="img/dynamic_acl_refresh.png" alt="Figure 2. Refresh dynamic ACL rule workflow">
+</p>
+
+#### Remove ACL rule when TTL expires
+<p align=center>
+<img src="img/dynamic_acl_expiration.png" alt="Figure 1. Create dynamic ACL rule workflow">
+</p>
+
+### Testing Requirements/Design  
+
+#### Unit Test cases  
+
+1. Enhance unit test for `acl-loader` to verify ACL rule with TTL is created as expected; verify `ACL_TTL_TABLE` entry is created as expected.
+2. Add unit test for `acl_ttl_checker` to verify expired ACL entries are cleared. 
+
+
+#### System Test cases
+
+1. Test case 1 Verify dynamic ACL rule is created as expected
+2. Test case 2 Verify dynamic ACL rule can be refreshed
+3. Test case 3 Verify expired dynamic ACL rule can be removed
```

Comment:
The second one `dynamic ACL should be removed from config_db after reboot` makes sense to me. I'll add a test case to verify dynamic ACL is cleared after config reload or reboot.
But why do we need the first verification `Convert static ACL rule to dynamic ACL rule is not allowed`?

Reply:
In case someone changed existing static rules to dynamic rules.
For dynamic ACL, the basic principle should be JIT change. The state of switch shouldn't be changed after dynamic ACL expired.

Consider below use case:
1. Someone pushed a dynamic rule that have exact same name with existing static rule
2. The rule will be removed after it expired, the switch no longer have that static rule anymore

We should prevent override static rule by pushing dynamic rule. Otherwise the dynamic rule will have side-effect after it expired.